docs/design/2022-04-21-performance-diagnosis-enhancements.md

```diff
@@ -0,0 +1,188 @@
+
+## Motivation or Background
+
+TiDB collects many metrics, but few of them are frequently used and the metrics of the critical paths are in different panels, which makes the diagnosis a problem. Furthermore, some latency events are not recorded in critical paths.
+
+This RPC proposes some improvements to the metrics collection, and display.
+
+## Detailed Design
+
+### Differences Between Internal and External Requests
+
+The tracking issue of it is #33963, and it should be applied to most of the metrics.
+
+#### Internal Transaction Usage
+
+The main goal of this task is to differentiate the internal and external transactions and mark the snapshots and transactions as internal for internal sessions and tasks.
+
+Some tasks are not running in the internal sessions, however, they are executed directly by `RunInNewTxn` :
+
+- DDL Backfill
+- Auto-ID management
+- Version check when bootstrapping
+- GC placement and label rules
+- Load data for cached tables
+- Alloc global ID for some operations
+    - Write sequence update to binlog
+    - Create temporary table
+
+When TiDB bootstrapping, it’ll create some internal sessions, these sessions are used for:
+
+- Bind info read/write
+- Privilege read/write
+- System variable read/write
+- Telemetry report
+- Statistic load
+
+The coprocessor request, transaction, and snapshot created by `RunInNewTxn` and the internal sessions need to be marked as internal.
+
+#### kvproto
+
+Add a field `request_source` in the proto, which is a string type field and can be empty(TiDB does not allow empty `request_source`). The string type can make the metrics display more readable.
+
+```diff
+// Miscellaneous metadata attached to most requests.
+message Context {
+    reserved 4;
+    ...
++   // The source of the request.
++   string request_source = xx;
+}
+
+```
+
+In TiDB, `request_source` is concatenated by **scope** and **type**, the scope is one of "internal" and "external", for type, there will be some variation between internal and external transactions.
+
+The type of the internal transactions would be their capabilities, the internal transactions are listed in the [Internal Transaction Usage](#internal-transaction-usage) section. The concatenated `request_source` would be like "internal_ddl_backfill_index", "internal_bindinfo_load", "internal_statistic_load", etc.
+
+The type of external transaction is the statement type. The concatenated `request_source` would be like "external_select", "external_insert", "external_commit", etc.
+
+#### Session
+
+Add a flag in `SessionVars`, TiDB creates some internal sessions when bootstrap, and the internal flags are set after bootstrap.
+
+```diff
+type SessionVars struct {
+	...
++	Internal bool
+}
+```
+
+#### Coprocessor
+
+```diff
+// Request represents a kv request.
+type Request struct {
+	...
++	RequestSourceScope string
```

Comment:
You already added 'string request_source = xx' to 'kvrpcpb.Context', and here is a field of 'kvrpcpb.Context' in the struct 'corprocessor.Request', why we need add these two 'RequestSourceScope' and 'RequestSourceType' to 'corprocessor.Request' ?

Reply:
The reason is the coprocessor requests are built bypassing kv client.
 
[`kv.Request`](https://github.com/pingcap/tidb/blob/65f329e78ea4fd7ab46b77db25eb1e72e6ca64cf/kv/kv.go#L308-L362) is a struct inside TiDB which will be converted to `kvrpcpb.Context` [here](https://github.com/pingcap/tidb/blob/65f329e78ea4fd7ab46b77db25eb1e72e6ca64cf/store/copr/coprocessor.go#L739-L746).